File: app/main/utils/quest.lua
local utils = require "main.utils.utils"
local npcs = require "main.states.npcstates"

local M = {}

function M.has_value(t)
	return utils.has_value(t[1](),t[2])
end

function M.any_has_value(t)
	local npcs= t[1]()
	for ni,nv in pairs(npcs) do 
		if utils.has_value(nv.inventory,t[2]) then return true end
	end
	return false
end

function M.returnfalse()
	return false
end

function M.convos_check(args)
	local npcs = args[1]()
	for nk,nv in pairs(npcs) do
		if nv.convos >= args[2] then return true end
	end
	return false
end

function M.max_skills(args)
	local skills = args[1]()
	local order = utils.create_ipairs(skills)
	table.sort(order, function(a, b) return skills[a] > skills[b] end)	
	for count = 1, args[2] do
		if skills[order[count]] < 10 then return false end
	end
	return true
end

function M.max_love(args)
	local count = 10
	local score = 0
	local all = args[1]()
	for _,nv in pairs(utils.shuffle(npcs.order)) do
		if all[nv].love > 5 then score=score+1 end
		count = count-1
		if score > args[2] then return true end
		if count == 0 then return false end
	end
end

return M

File: src/scripts/utils/quest.ts
import { Npcs, Skills } from '../../types/state'
import NpcState from '../states/npc'
//import { shuffle } from './utils'

export function has_value(t: [() => string[], string]) {
  return t[0]().includes(t[1])
}

export function any_has_value(t: [() => NpcState[], string]) {
  const npcs = t[0]()
  for (const npc of npcs) {
    if (npc.inventory.includes(t[1])) return true
  }
  return false
}

export function is_same_room(t: [() => NpcState[], string]) {
  const npcs = t[0]()
  for (const npc of npcs) {
    if (npc.currentroom == t[1]) return true
  }
  return false
}

export function from_same_room(npcs: NpcState[], room: string) {
  for (const npc of npcs) {
    if (npc.currentroom == room) return npc
  }
  return null
}

export function returnfalse() {
  return false
}

export function does_equal(
  args: [() => number | string, number | string]
): boolean {
  return args[0]() == args[1] ? true : false
}

export function greater_than(args: [() => number, number]): boolean {
  const value = args[0]()
  return value > args[1] ? true : false
}

export function convos_check(args: [() => Npcs, number]): boolean {
  const npcs = args[0]()
  let nKey: keyof typeof npcs
  for (nKey in npcs) {
    if (npcs[nKey].convos >= args[1]) {
      return true //tesjpf could return array with "hounded" ??? bad.
    }
  }
  return false
}

export function max_skills(args: [() => Skills, number]) {
  const skills = args[0]()
  const order = Object.entries(skills)

  order.sort((a: [string, number], b: [string, number]) => b[1] - a[1])

  for (let i = args[1]; i-- !== 0; ) {
    if (order[i][1] < 10) {
      return false
    }
  }
  return true
}

export function max_love(args: [() => [string[], Npcs], number]): boolean {
  let count = 10
  let score = 0
  const [ordered, all] = args[0]()
  for (const npcname of ordered) {
    if (all[npcname].love > args[1]) {
      score = score + 1
    }
    count = count - 1
    if (score > args[1]) return true
    if (count == 0) return false
  }
  return false
}
